include/pybind11/numpy.h

```diff
@@ -250,6 +267,64 @@ struct npy_api {
     }
 };
 
+template <typename T> struct is_complex : std::false_type { };
+template <typename T> struct is_complex<std::complex<T>> : std::true_type { };
+
+template <typename T, typename = void>
+struct npy_format_descriptor_name;
+
+template <typename T>
+struct npy_format_descriptor_name<T, enable_if_t<std::is_integral<T>::value>> {
+    static constexpr auto name = _<std::is_same<T, bool>::value>(
+        _("bool"), _<std::is_signed<T>::value>("int", "uint") + _<sizeof(T)*8>()
+    );
+};
+
+template <typename T>
+struct npy_format_descriptor_name<T, enable_if_t<std::is_floating_point<T>::value>> {
+    static constexpr auto name = _<std::is_same<T, float>::value || std::is_same<T, double>::value>(
+        _("float") + _<sizeof(T)*8>(), _("longdouble")
+    );
+};
+
+template <typename T>
+struct npy_format_descriptor_name<T, enable_if_t<is_complex<T>::value>> {
+    static constexpr auto name = _<std::is_same<typename T::value_type, float>::value
+                                   || std::is_same<typename T::value_type, double>::value>(
+        _("complex") + _<sizeof(typename T::value_type)*16>(), _("longcomplex")
+    );
+};
+
+template<typename T> struct numpy_scalar_info {};
+
+#define DECL_NPY_SCALAR(ctype_, typenum_) \
+    template<> struct numpy_scalar_info<ctype_> { \
+        static constexpr auto name = npy_format_descriptor_name<ctype_>::name; \
+        static constexpr int typenum = npy_api::typenum_##_; \
+    }
+
+DECL_NPY_SCALAR(bool, NPY_BOOL);
+
+DECL_NPY_SCALAR(int8_t, NPY_INT8);
```

Comment:
I am not entirely sure if specializing for fundamental types `short`, `int`, `long` and `long long` isn't the better choice to cover all types (and then mapping to concrete bit representations).

Reply:
Yea, I'm not sure about that either, it's all a bit dodgy. But I guess if done this way it's at least guaranteed they don't overlap.

Reply:
Yes, but it does not guarantee that on platforms where `int` is implemented with the same size as `long` both types will be covered by the API.

Reply:
As I've said, I don't have very strong preference, it's somewhat dodgy either way, but we could do something like this (and hope that the stars align in the right direction):

```c++
DECL_NPY_SCALAR(signed char, NPY_BYTE);
DECL_NPY_SCALAR(short, NPY_SHORT);
DECL_NPY_SCALAR(int, NPY_INT);
DECL_NPY_SCALAR(long, NPY_LONG);
DECL_NPY_SCALAR(long long, NPY_LONGLONG);

DECL_NPY_SCALAR(unsigned char, NPY_UBYTE);
DECL_NPY_SCALAR(unsigned short, NPY_USHORT);
DECL_NPY_SCALAR(unsigned int, NPY_UINT);
DECL_NPY_SCALAR(unsigned long, NPY_ULONG);
DECL_NPY_SCALAR(unsigned long long, NPY_ULONGLONG);
```

Because there's 5 of each signed/unsigned standard types in C, there will be guaranteed to be an overlap (i.e. almost always `long` and `long long`). But I guess that's the way you have to do it "by the book".

Unless I'm missing something, that's probably ok, we will have 10 type casters for 10 distinct C types which will map to 8 (or less) distinct numpy types.

Reply:
Technically, there's another special type here - `char`, which may be either `NPY_BYTE` or `NPY_UBYTE` depending on signedness.

Anyways, I've update the mappings and added the `char` as well + a note in the doc.

Reply:
actually there are even three *distinct* char types: `char`, `signed char` and `unsigned char` ^^
https://en.cppreference.com/w/cpp/language/types#Character_types

Reply:
> actually there are even three distinct char types: unsigned char, signed char and char ^^

Have you read the last diff? :)